File: lib/pricing-engine.ts
/**
 * Pricing Engine
 * Deterministic financial calculations for EPC bid pricing.
 *
 * NO LLM IS USED HERE. Per PRD: math must be deterministic and auditable.
 * Contract terms (LD rate, retention %, security %) are extracted from the
 * source document via regex, falling back to industry-standard defaults
 * when a term isn't explicitly stated.
 */

export interface PricingInputs {
  contract_value: number;
  estimated_duration_weeks: number;
  material_cost_percentage: number; // fraction of contract value spent on materials
  labor_cost_percentage: number; // fraction of contract value spent on labor
  contingency_percentage: number; // fraction added as risk buffer
  target_margin_percentage: number; // desired profit margin on bid price
  ld_rate_per_week_percentage: number; // % of contract value deducted per week of delay
  ld_cap_percentage: number; // max % of contract value exposed to LD
  performance_security_percentage: number; // % of contract value held as performance bond
  retention_rate_per_invoice_percentage: number; // % withheld from each invoice
  retention_cap_percentage: number; // max % of contract value withheld as retention
  invoices_count: number; // number of progress invoices expected over the project
}

export interface PricingBreakdown {
  contract_value: number;

  // Cost buildup
  material_cost: number;
  labor_cost: number;
  contingency_amount: number;
  total_estimated_cost: number;

  // Bid pricing
  target_margin_percentage: number;
  recommended_bid_price: number;

  // Liquidated damages
  ld_rate_per_week_percentage: number;
  ld_exposure_per_week: number;
  ld_cap_percentage: number;
  ld_cap_amount: number;
  weeks_to_reach_cap: number;

  // Performance security
  performance_security_percentage: number;
  performance_security_amount: number;

  // Retention
  estimated_invoice_amount: number;
  retention_rate_per_invoice_percentage: number;
  retention_per_invoice: number;
  retention_cap_percentage: number;
  retention_cap_amount: number;
  total_retention_held: number;

  // Cash flow / lock-up
  total_lockup: number;
  lockup_as_percentage_of_contract: number;

  // Assumptions used (for auditability)
  assumptions_used: string[];
}

/**
 * Industry-standard defaults, used only when a term cannot be extracted
 * from the source document.
 */
export const DEFAULT_PRICING_PARAMETERS = {
  material_cost_percentage: 0.5,
  labor_cost_percentage: 0.3,
  contingency_percentage: 0.1,
  target_margin_percentage: 0.15,
  ld_rate_per_week_percentage: 0.01, // 1% per week
  ld_cap_percentage: 0.1, // capped at 10%
  performance_security_percentage: 0.1, // 10%
  retention_rate_per_invoice_percentage: 0.1, // 10% per RA bill
  retention_cap_percentage: 0.05, // capped at 5%
  invoices_count: 12,
};

/**
 * Core deterministic pricing calculation.
 * No LLM calls - pure math per PRD requirement.
 */
export function calculatePricing(inputs: PricingInputs): PricingBreakdown {
  const {
    contract_value,
    estimated_duration_weeks,
    material_cost_percentage,
    labor_cost_percentage,
    contingency_percentage,
    target_margin_percentage,
    ld_rate_per_week_percentage,
    ld_cap_percentage,
    performance_security_percentage,
    retention_rate_per_invoice_percentage,
    retention_cap_percentage,
    invoices_count,
  } = inputs;

  // --- Cost buildup ---
  const material_cost = round2(contract_value * material_cost_percentage);
  const labor_cost = round2(contract_value * labor_cost_percentage);
  const base_cost = material_cost + labor_cost;
  const contingency_amount = round2(base_cost * contingency_percentage);
  const total_estimated_cost = round2(base_cost + contingency_amount);

  // --- Bid pricing ---
  const recommended_bid_price = round2(
    total_estimated_cost * (1 + target_margin_percentage),
  );

  // --- Liquidated damages ---
  const ld_exposure_per_week = round2(contract_value * ld_rate_per_week_percentage);
  const ld_cap_amount = round2(contract_value * ld_cap_percentage);
  const weeks_to_reach_cap =
    ld_exposure_per_week > 0
      ? Math.ceil(ld_cap_amount / ld_exposure_per_week)
      : Infinity;

  // --- Performance security ---
  const performance_security_amount = round2(
    contract_value * performance_security_percentage,
  );

  // --- Retention ---
  // Compute on unrounded values to avoid compounding rounding error across
  // `invoices_count` installments; round only once, for the returned totals.
  const raw_invoice_amount = invoices_count > 0 ? contract_value / invoices_count : 0;
  const raw_retention_per_invoice = raw_invoice_amount * retention_rate_per_invoice_percentage;
  const estimated_invoice_amount = round2(raw_invoice_amount);
  const retention_per_invoice = round2(raw_retention_per_invoice);
  const retention_cap_amount = round2(contract_value * retention_cap_percentage);
  const total_retention_held = Math.min(
    round2(raw_retention_per_invoice * invoices_count),
    retention_cap_amount,
  );

  // --- Total lock-up (cash tied up during project execution) ---
  const total_lockup = round2(performance_security_amount + total_retention_held);
  const lockup_as_percentage_of_contract = round4(
    contract_value > 0 ? total_lockup / contract_value : 0,
  );

  return {
    contract_value,
    material_cost,
    labor_cost,
    contingency_amount,
    total_estimated_cost,
    target_margin_percentage,
    recommended_bid_price,
    ld_rate_per_week_percentage,
    ld_exposure_per_week,
    ld_cap_percentage,
    ld_cap_amount,
    weeks_to_reach_cap,
    performance_security_percentage,
    performance_security_amount,
    estimated_invoice_amount,
    retention_rate_per_invoice_percentage,
    retention_per_invoice,
    retention_cap_percentage,
    retention_cap_amount,
    total_retention_held,
    total_lockup,
    lockup_as_percentage_of_contract,
    assumptions_used: [],
  };
}

/**
 * Extracts pricing parameters from raw document text using regex.
 * Falls back to DEFAULT_PRICING_PARAMETERS for anything not found.
 * Returns the resolved inputs plus a human-readable list of which
 * values were extracted vs. defaulted (for auditability).
 */
export function extractPricingInputs(
  documentText: string,
  estimatedDurationWeeksFallback: number = 24,
): { inputs: PricingInputs; assumptionsUsed: string[] } {
  const assumptionsUsed: string[] = [];

  const contract_value = extractContractValue(documentText, assumptionsUsed);
  const estimated_duration_weeks = extractDurationWeeks(
    documentText,
    estimatedDurationWeeksFallback,
    assumptionsUsed,
  );

  const ld_rate_per_week_percentage = extractPercentage(
    documentText,
    /(\d+(?:\.\d+)?)\s*%\s*(?:of\s*(?:the\s*)?contract\s*value\s*)?per\s*week(?:\s*(?:of\s*)?delay)?/i,
    DEFAULT_PRICING_PARAMETERS.ld_rate_per_week_percentage,
    'ld_rate_per_week_percentage',
    assumptionsUsed,
  );

  const ld_cap_percentage = extractPercentage(
    documentText,
    /(?:liquidated\s*damages|delay\s*penalt(?:y|ies))[\s\S]{0,120}?capped\s*at\s*(\d+(?:\.\d+)?)\s*%/i,
    DEFAULT_PRICING_PARAMETERS.ld_cap_percentage,
    'ld_cap_percentage',
    assumptionsUsed,
  );

  const performance_security_percentage = extractPercentage(
    documentText,
    /performance\s*(?:bond|security)[:\s]*(?:is\s*)?(\d+(?:\.\d+)?)\s*%/i,
    DEFAULT_PRICING_PARAMETERS.performance_security_percentage,
    'performance_security_percentage',
    assumptionsUsed,
  );

  const retention_rate_per_invoice_percentage = extractPercentage(
    documentText,
    /retention[:\s]*(?:is\s*)?(\d+(?:\.\d+)?)\s*%/i,
    DEFAULT_PRICING_PARAMETERS.retention_rate_per_invoice_percentage,
    'retention_rate_per_invoice_percentage',
    assumptionsUsed,
  );

  const retention_cap_percentage = extractPercentage(
    documentText,
    /retention[\s\S]{0,80}?capped\s*at\s*(\d+(?:\.\d+)?)\s*%/i,
    DEFAULT_PRICING_PARAMETERS.retention_cap_percentage,
    'retention_cap_percentage',
    assumptionsUsed,
  );

  const inputs: PricingInputs = {
    contract_value,
    estimated_duration_weeks,
    material_cost_percentage: DEFAULT_PRICING_PARAMETERS.material_cost_percentage,
    labor_cost_percentage: DEFAULT_PRICING_PARAMETERS.labor_cost_percentage,
    contingency_percentage: DEFAULT_PRICING_PARAMETERS.contingency_percentage,
    target_margin_percentage: DEFAULT_PRICING_PARAMETERS.target_margin_percentage,
    ld_rate_per_week_percentage,
    ld_cap_percentage,
    performance_security_percentage,
    retention_rate_per_invoice_percentage,
    retention_cap_percentage,
    invoices_count: DEFAULT_PRICING_PARAMETERS.invoices_count,
  };

  return { inputs, assumptionsUsed };
}

/**
 * Convenience wrapper: extract inputs from document text and calculate
 * pricing in one call. Attaches the assumptions list to the result.
 */
export function calculatePricingFromDocument(
  documentText: string,
  estimatedDurationWeeksFallback: number = 24,
): PricingBreakdown {
  const { inputs, assumptionsUsed } = extractPricingInputs(
    documentText,
    estimatedDurationWeeksFallback,
  );

  const breakdown = calculatePricing(inputs);
  breakdown.assumptions_used = assumptionsUsed;
  return breakdown;
}

// ---------------------------------------------------------------------------
// Extraction helpers
// ---------------------------------------------------------------------------

function extractContractValue(text: string, assumptionsUsed: string[]): number {
  // Matches: "Contract Value: USD 12,500,000" / "Project Value: USD 12.5 Million"
  const millionMatch = text.match(
    /(?:contract|project)\s*value[:\s]*(?:USD|US\$|\$)?\s*([\d,]+(?:\.\d+)?)\s*million/i,
  );
  if (millionMatch) {
    return parseFloat(millionMatch[1].replace(/,/g, '')) * 1_000_000;
  }

  const exactMatch = text.match(
    /(?:contract|project)\s*value[:\s]*(?:USD|US\$|\$)?\s*([\d,]+(?:\.\d+)?)/i,
  );
  if (exactMatch) {
    return parseFloat(exactMatch[1].replace(/,/g, ''));
  }

  assumptionsUsed.push(
    'contract_value: not found in document, defaulted to 0 - caller must supply explicitly',
  );
  return 0;
}

function extractDurationWeeks(
  text: string,
  fallback: number,
  assumptionsUsed: string[],
): number {
  const monthsMatch = text.match(/duration[:\s]*(\d+)\s*months?/i);
  if (monthsMatch) {
    return Math.round(parseInt(monthsMatch[1], 10) * 4.345);
  }

  const weeksMatch = text.match(/duration[:\s]*(\d+)\s*weeks?/i);
  if (weeksMatch) {
    return parseInt(weeksMatch[1], 10);
  }

  assumptionsUsed.push(
    `estimated_duration_weeks: not found in document, defaulted to ${fallback} weeks`,
  );
  return fallback;
}

function extractPercentage(
  text: string,
  regex: RegExp,
  fallback: number,
  fieldName: string,
  assumptionsUsed: string[],
): number {
  const match = text.match(regex);
  if (match) {
    return parseFloat(match[1]) / 100;
  }

  assumptionsUsed.push(
    `${fieldName}: not found in document, defaulted to ${(fallback * 100).toFixed(1)}%`,
  );
  return fallback;
}

function round2(value: number): number {
  return Math.round(value * 100) / 100;
}

function round4(value: number): number {
  return Math.round(value * 10000) / 10000;
}

File: python/app/pricing_engine.py
"""Deterministic EPC pricing calculations. Ported from lib/pricing-engine.ts.
No LLM calls - pure math, contract terms regex-extracted with defaults."""

from __future__ import annotations

import math
import re
from typing import Any

DEFAULT_PRICING_PARAMETERS = {
    "material_cost_percentage": 0.5,
    "labor_cost_percentage": 0.3,
    "contingency_percentage": 0.1,
    "target_margin_percentage": 0.15,
    "ld_rate_per_week_percentage": 0.01,
    "ld_cap_percentage": 0.1,
    "performance_security_percentage": 0.1,
    "retention_rate_per_invoice_percentage": 0.1,
    "retention_cap_percentage": 0.05,
    "invoices_count": 12,
}


def _round2(value: float) -> float:
    return round(value * 100) / 100


def _round4(value: float) -> float:
    return round(value * 10000) / 10000


def calculate_pricing(inputs: dict[str, Any]) -> dict[str, Any]:
    contract_value = inputs["contract_value"]
    estimated_duration_weeks = inputs["estimated_duration_weeks"]
    material_cost_percentage = inputs["material_cost_percentage"]
    labor_cost_percentage = inputs["labor_cost_percentage"]
    contingency_percentage = inputs["contingency_percentage"]
    target_margin_percentage = inputs["target_margin_percentage"]
    ld_rate_per_week_percentage = inputs["ld_rate_per_week_percentage"]
    ld_cap_percentage = inputs["ld_cap_percentage"]
    performance_security_percentage = inputs["performance_security_percentage"]
    retention_rate_per_invoice_percentage = inputs["retention_rate_per_invoice_percentage"]
    retention_cap_percentage = inputs["retention_cap_percentage"]
    invoices_count = inputs["invoices_count"]

    material_cost = _round2(contract_value * material_cost_percentage)
    labor_cost = _round2(contract_value * labor_cost_percentage)
    base_cost = material_cost + labor_cost
    contingency_amount = _round2(base_cost * contingency_percentage)
    total_estimated_cost = _round2(base_cost + contingency_amount)

    recommended_bid_price = _round2(total_estimated_cost * (1 + target_margin_percentage))

    ld_exposure_per_week = _round2(contract_value * ld_rate_per_week_percentage)
    ld_cap_amount = _round2(contract_value * ld_cap_percentage)
    # None (not math.inf) when there's no weekly LD exposure to divide by -
    # math.inf serializes to the bare token `Infinity`, which is invalid
    # JSON and breaks writing this dict to a Postgres JSONB column.
    weeks_to_reach_cap = (
        math.ceil(ld_cap_amount / ld_exposure_per_week) if ld_exposure_per_week > 0 else None
    )

    performance_security_amount = _round2(contract_value * performance_security_percentage)

    raw_invoice_amount = contract_value / invoices_count if invoices_count > 0 else 0
    raw_retention_per_invoice = raw_invoice_amount * retention_rate_per_invoice_percentage
    estimated_invoice_amount = _round2(raw_invoice_amount)
    retention_per_invoice = _round2(raw_retention_per_invoice)
    retention_cap_amount = _round2(contract_value * retention_cap_percentage)
    total_retention_held = min(_round2(raw_retention_per_invoice * invoices_count), retention_cap_amount)

    total_lockup = _round2(performance_security_amount + total_retention_held)
    lockup_as_percentage_of_contract = _round4(total_lockup / contract_value if contract_value > 0 else 0)

    return {
        "contract_value": contract_value,
        "material_cost": material_cost,
        "labor_cost": labor_cost,
        "contingency_amount": contingency_amount,
        "total_estimated_cost": total_estimated_cost,
        "target_margin_percentage": target_margin_percentage,
        "recommended_bid_price": recommended_bid_price,
        "ld_rate_per_week_percentage": ld_rate_per_week_percentage,
        "ld_exposure_per_week": ld_exposure_per_week,
        "ld_cap_percentage": ld_cap_percentage,
        "ld_cap_amount": ld_cap_amount,
        "weeks_to_reach_cap": weeks_to_reach_cap,
        "performance_security_percentage": performance_security_percentage,
        "performance_security_amount": performance_security_amount,
        "estimated_invoice_amount": estimated_invoice_amount,
        "retention_rate_per_invoice_percentage": retention_rate_per_invoice_percentage,
        "retention_per_invoice": retention_per_invoice,
        "retention_cap_percentage": retention_cap_percentage,
        "retention_cap_amount": retention_cap_amount,
        "total_retention_held": total_retention_held,
        "total_lockup": total_lockup,
        "lockup_as_percentage_of_contract": lockup_as_percentage_of_contract,
        "assumptions_used": [],
        "estimated_duration_weeks": estimated_duration_weeks,
    }


def _extract_contract_value(text: str, assumptions_used: list[str]) -> float:
    million_match = re.search(
        r"(?:contract|project)\s*value[:\s]*(?:USD|US\$|\$)?\s*(\d[\d,]*(?:\.\d+)?)\s*million",
        text,
        re.IGNORECASE,
    )
    if million_match:
        return float(million_match.group(1).replace(",", "")) * 1_000_000

    exact_match = re.search(
        r"(?:contract|project)\s*value[:\s]*(?:USD|US\$|\$)?\s*(\d[\d,]*(?:\.\d+)?)", text, re.IGNORECASE
    )
    if exact_match:
        return float(exact_match.group(1).replace(",", ""))

    assumptions_used.append(
        "contract_value: not found in document, defaulted to 0 - caller must supply explicitly"
    )
    return 0


def _extract_duration_weeks(text: str, fallback: int, assumptions_used: list[str]) -> int:
    months_match = re.search(r"duration[:\s]*(\d+)\s*months?", text, re.IGNORECASE)
    if months_match:
        return round(int(months_match.group(1)) * 4.345)

    weeks_match = re.search(r"duration[:\s]*(\d+)\s*weeks?", text, re.IGNORECASE)
    if weeks_match:
        return int(weeks_match.group(1))

    assumptions_used.append(f"estimated_duration_weeks: not found in document, defaulted to {fallback} weeks")
    return fallback


def _extract_percentage(
    text: str, pattern: str, fallback: float, field_name: str, assumptions_used: list[str]
) -> float:
    match = re.search(pattern, text, re.IGNORECASE)
    if match:
        return float(match.group(1)) / 100

    assumptions_used.append(f"{field_name}: not found in document, defaulted to {fallback * 100:.1f}%")
    return fallback


def extract_pricing_inputs(
    document_text: str, estimated_duration_weeks_fallback: int = 24
) -> tuple[dict[str, Any], list[str]]:
    assumptions_used: list[str] = []

    contract_value = _extract_contract_value(document_text, assumptions_used)
    estimated_duration_weeks = _extract_duration_weeks(
        document_text, estimated_duration_weeks_fallback, assumptions_used
    )

    ld_rate_per_week_percentage = _extract_percentage(
        document_text,
        r"(\d+(?:\.\d+)?)\s*%\s*(?:of\s*(?:the\s*)?contract\s*value\s*)?per\s*week(?:\s*(?:of\s*)?delay)?",
        DEFAULT_PRICING_PARAMETERS["ld_rate_per_week_percentage"],
        "ld_rate_per_week_percentage",
        assumptions_used,
    )
    ld_cap_percentage = _extract_percentage(
        document_text,
        r"(?:liquidated\s*damages|delay\s*penalt(?:y|ies))[\s\S]{0,120}?capped\s*at\s*(\d+(?:\.\d+)?)\s*%",
        DEFAULT_PRICING_PARAMETERS["ld_cap_percentage"],
        "ld_cap_percentage",
        assumptions_used,
    )
    performance_security_percentage = _extract_percentage(
        document_text,
        r"performance\s*(?:bond|security)[:\s]*(?:is\s*)?(\d+(?:\.\d+)?)\s*%",
        DEFAULT_PRICING_PARAMETERS["performance_security_percentage"],
        "performance_security_percentage",
        assumptions_used,
    )
    retention_rate_per_invoice_percentage = _extract_percentage(
        document_text,
        r"retention[:\s]*(?:is\s*)?(\d+(?:\.\d+)?)\s*%",
        DEFAULT_PRICING_PARAMETERS["retention_rate_per_invoice_percentage"],
        "retention_rate_per_invoice_percentage",
        assumptions_used,
    )
    retention_cap_percentage = _extract_percentage(
        document_text,
        r"retention[\s\S]{0,80}?capped\s*at\s*(\d+(?:\.\d+)?)\s*%",
        DEFAULT_PRICING_PARAMETERS["retention_cap_percentage"],
        "retention_cap_percentage",
        assumptions_used,
    )

    inputs = {
        "contract_value": contract_value,
        "estimated_duration_weeks": estimated_duration_weeks,
        "material_cost_percentage": DEFAULT_PRICING_PARAMETERS["material_cost_percentage"],
        "labor_cost_percentage": DEFAULT_PRICING_PARAMETERS["labor_cost_percentage"],
        "contingency_percentage": DEFAULT_PRICING_PARAMETERS["contingency_percentage"],
        "target_margin_percentage": DEFAULT_PRICING_PARAMETERS["target_margin_percentage"],
        "ld_rate_per_week_percentage": ld_rate_per_week_percentage,
        "ld_cap_percentage": ld_cap_percentage,
        "performance_security_percentage": performance_security_percentage,
        "retention_rate_per_invoice_percentage": retention_rate_per_invoice_percentage,
        "retention_cap_percentage": retention_cap_percentage,
        "invoices_count": DEFAULT_PRICING_PARAMETERS["invoices_count"],
    }

    return inputs, assumptions_used


def calculate_pricing_from_document(document_text: str, estimated_duration_weeks_fallback: int = 24) -> dict[str, Any]:
    inputs, assumptions_used = extract_pricing_inputs(document_text, estimated_duration_weeks_fallback)
    breakdown = calculate_pricing(inputs)
    breakdown["assumptions_used"] = assumptions_used
    return breakdown
